validation/src/blockchain/eth_rpc.rs

```diff
@@ -0,0 +1,295 @@
+pub struct EthBlock {
+    hash: String,
+    number: String,
+    timestamp: String,
+}
+
+impl TryFrom<EthBlock> for BlockResponse {
+    type Error = anyhow::Error;
+
+    fn try_from(value: EthBlock) -> std::result::Result<Self, Self::Error> {
+        Ok(BlockResponse {
+            hash: value.hash,
+            number: i64_from_hex(&value.number).context("invalid block number")?,
+            timestamp: i64_from_hex(&value.timestamp).context("invalid block timestamp")?,
+        })
+    }
+}
+
+#[derive(Debug)]
+/// Http client to interact with EIP chains
+pub struct HttpEthRpc {
+    chain_id: caip2::ChainId,
+    url: reqwest::Url,
+    tx_cache: Arc<Mutex<LruCache<String, EthTransaction>>>,
+    block_cache: Arc<Mutex<LruCache<String, EthBlock>>>,
+}
+
+impl HttpEthRpc {
+    /// Create a new ethereum VM compatible HTTP client
+    pub async fn try_new(url: &str) -> Result<Self> {
+        let url = reqwest::Url::parse(url).context("invalid url")?;
+        let chain_id = Self::eth_chain_id(url.clone()).await?;
+        let chain_decimal = i64_from_hex(&chain_id)?;
+        let chain_id = caip2::ChainId::from_str(&format!("eip155:{chain_decimal}"))?;
+        let tx_cache = Arc::new(Mutex::new(LruCache::new(
+            NonZero::new(TRANSACTION_CACHE_SIZE).expect("transaction cache size must be non zero"),
+        )));
+        let block_cache = Arc::new(Mutex::new(LruCache::new(
+            NonZero::new(BLOCK_CACHE_SIZE).expect("block cache size must be non zero"),
+        )));
+        Ok(Self {
+            url,
+            chain_id,
+            tx_cache,
+            block_cache,
+        })
+    }
+
+    /// Retrieve the caip2 Chain ID
+    ///
+    /// ❯ curl https://cloudflare-eth.com/v1/mainnet -X POST --data '{"jsonrpc":"2.0","method":"eth_chainId","params":[],"id":0}'
+    ///    {"jsonrpc":"2.0","result":"0x1","id":0}
+    ///
+    /// ❯ curl https://rpc.ankr.com/filecoin -X POST --data '{"jsonrpc":"2.0","method":"eth_chainId","params":[],"id":0}'
+    /// >>  {"id":0,"jsonrpc":"2.0","result":"0x13a"}
+    async fn eth_chain_id(url: reqwest::Url) -> Result<String> {
+        let resp: RpcResponse<String> = HTTP_CLIENT
+            .post(url.clone())
+            .json(&serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_chainId",
+                "params": [],
+                "id": 0,
+            }))
+            .send()
+            .await?
+            .json()
+            .await?;
+        resp.result
+            .ok_or_else(|| anyhow!("failed to retrieve chain ID"))
+    }
+
+    /// Get the latest block number
+    ///
+    /// curl https://mainnet.infura.io/v3/{api_token} \
+    /// -X POST \
+    /// -H "Content-Type: application/json" \
+    /// -d '{"jsonrpc":"2.0","method":"eth_blockNumber","params": [],"id":1}'
+    /// >> {"jsonrpc": "2.0", "id": 1, "result": "0x127cc18"}
+    async fn _eth_block_number(&self) -> Result<RpcResponse<String>> {
```

Comment:
is this used anywhere?  Do we need it?

Reply:
no, probably not... it was originally used to make sure we had mined "enough blocks" to consider the timestamp valid.. I left if in here in case we want to do that but it's simple to add back later.